src/platform/datapath_epoll.c

```diff
@@ -1644,10 +1499,273 @@ CxPlatSocketContextSendComplete(
+
+    BOOLEAN UseURO = !!(SocketContext->Binding->Datapath->Features & CXPLAT_DATAPATH_FEATURE_RECV_COALESCING);
+    int BufferCount = UseURO ? 1 : NUMBER_OF_SEGMENTS_OR_BATCHES;
+    struct iovec* RecvIov = RecvMessageStore->RecvIov;
+    CXPLAT_RECV_MSG_CONTROL_BUFFER* RecvMsgControl = RecvMessageStore->RecvMsgControl;
+    struct mmsghdr* RecvMsgHdr = RecvMessageStore->RecvMsgHdr;
+
+    if (UseURO) {
+        //
+        // Configure as 1 URO buffer
+        //
+        RecvIov[0].iov_base = RecvPayload;
+        RecvIov[0].iov_len = NUMBER_OF_SEGMENTS_OR_BATCHES * MAX_GRO_SEGMENT_SIZE;
+    } else {
+        //
+        // Configure as 64 smaller buffers
+        //
+        for (int i = 0; i < NUMBER_OF_SEGMENTS_OR_BATCHES; i++) {
+            RecvIov[i].iov_base = (i * MAX_GRO_SEGMENT_SIZE) + RecvPayload;
+            RecvIov[i].iov_len = MAX_GRO_SEGMENT_SIZE;
+        }
+    }
+
+    CXPLAT_RECV_DATA* Datagram = (CXPLAT_RECV_DATA*)(RecvContext + 1);
+    CXPLAT_DATAPATH_INTERNAL_RECV_BUFFER_CONTEXT* InternalDatagramContext;
+
+    for (int i = 0; i < BufferCount; i++) {
+        struct msghdr* MsgHdr = &RecvMsgHdr[i].msg_hdr;
+        InternalDatagramContext =
+            CxPlatDataPathDatagramToInternalDatagramContext(Datagram);
+
+        MsgHdr->msg_name = &InternalDatagramContext->Tuple.RemoteAddress;
+        MsgHdr->msg_namelen = sizeof(InternalDatagramContext->Tuple.RemoteAddress);
+        MsgHdr->msg_iov = &RecvIov[i];
+        MsgHdr->msg_iovlen = 1;
+        MsgHdr->msg_control = &RecvMsgControl[i].Data;
+        MsgHdr->msg_controllen = sizeof(RecvMsgControl[i].Data);
+        MsgHdr->msg_flags = 0;
+
+        Datagram = (CXPLAT_RECV_DATA*)(((uint8_t*)Datagram) + Datapath->DatagramStride);
+    }
+
+    int Ret =
+        recvmmsg(
+            SocketContext->SocketFd,
+            RecvMsgHdr,
+            BufferCount,
+            0,
+            NULL);
+    if (Ret < 0) {
+        int Error = errno;
+        if (Error != EAGAIN && Error != EWOULDBLOCK) {
+            Status = Error;
+            QuicTraceEvent(
+                DatapathErrorStatus,
+                "[data][%p] ERROR, %u, %s.",
+                SocketContext->Binding,
+                Error,
+                "recvmmsg failed");
+        } else {
+            Status = QUIC_STATUS_PENDING;
+        }
+        goto Exit;
+    }
+
+    struct mmsghdr* Msg = &RecvMsgHdr[0];
+    Datagram = (CXPLAT_RECV_DATA*)(RecvContext + 1);
+    InternalDatagramContext =
+            CxPlatDataPathDatagramToInternalDatagramContext(Datagram);
+
+    QUIC_ADDR* LocalAddr = &InternalDatagramContext->Tuple.LocalAddress;
+    QUIC_ADDR* RemoteAddr = &InternalDatagramContext->Tuple.RemoteAddress;
+    uint8_t ECN;
+    uint16_t SegmentSize = 0;
+
+    if (!CxPlatSocketReadCmsg(&Msg->msg_hdr, SocketContext->Binding, LocalAddr, &ECN, &SegmentSize)) {
+        // TODO Error case missing ECN or Local Addr
+        CXPLAT_FRE_ASSERT(FALSE);
```

Comment:
Should we really fre assert?

Reply:
No, that's why it has the giant todo on it. I wanted to make sure it worked then finish cleaning up.